src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java

```diff
@@ -113,16 +118,32 @@ public static boolean useItem(ItemStack itemStack, EntityPlayer player, int coun
     }
 
     public static int countItem(ItemStack itemStack, EntityPlayer player, World world) {
+        return countItem(itemStack, player, world, null);
+    }
+
+    public static int countItem(ItemStack itemStack, EntityPlayer player, RemoteInventoryCache cache) {
+        return countItem(itemStack, player, null, cache);
+    }
+
+    private static int countItem(ItemStack itemStack, EntityPlayer player, @Nullable World world, @Nullable RemoteInventoryCache cache) {
         if (player.capabilities.isCreativeMode) {
             return 10000;
         }
         int count = 0;
         ItemStack tool = GadgetGeneric.getGadget(player);
-        BlockPos tePos = GadgetUtils.getBoundTE(tool, world);
-        if (tePos != null) {
-            TileEntity te = world.getTileEntity(tePos);
-            IItemHandler cap = te.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
-            count += countInContainer(cap, itemStack.getItem(), itemStack.getMetadata());
+        if (cache != null) {
+            Multiset<UniqueItem> cacheInv = cache.getCache(tool);
+            if (cacheInv != null)
+                count += cacheInv.count(new UniqueItem(itemStack.getItem(), itemStack.getMetadata()));
+        } else if (world != null) {
+            BlockPos tePos = GadgetUtils.getBoundTE(tool, world);
+            if (tePos != null) {
+                TileEntity te = world.getTileEntity(tePos);
+                if (te != null) {
+                    IItemHandler cap = te.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
+                    count += countInContainer(cap, itemStack.getItem(), itemStack.getMetadata());
+                }
+            }
```

Comment:
The interface approach would make this a lot nicer...

Reply:
Implemented.